src/Arcus.Scripting.ActiveDirectory/Scripts/Add-AzADAppRoleAssignment.ps1

```diff
@@ -0,0 +1,71 @@
+)
+
+$adApplication = Get-AzADApplication -Filter "AppId eq '$ClientId'"
+if (!$adApplication) { 
+    throw "Active Directory Application for the ClientId '$ClientId' could not be found" 
+}
+$adServicePrincipal = Get-AzADServicePrincipal -Filter "AppId eq '$ClientId'"
+if (!$adServicePrincipal) { 
+    throw "Active Directory Service Principal for the ClientId '$ClientId' could not be found" 
+}
+
+$adApplicationRoleAssignTo = Get-AzADApplication -Filter "AppId eq '$AssignRoleToClientId'"
+if (!$adApplicationRoleAssignTo) { 
+    throw "Active Directory Application for the ClientId '$AssignRoleToClientId' could not be found" 
+}
+$adServicePrincipalRoleAssignTo = Get-AzADServicePrincipal -Filter "AppId eq '$AssignRoleToClientId'"
+if (!$adServicePrincipalRoleAssignTo) { 
+    throw "Active Directory Service Principal for the ClientId '$AssignRoleToClientId' could not be found" 
+}
+
+try {
+    if (-not ($adApplication.AppRole.Value -contains $Role)) {
+        Write-Host "Active Directory Application '$($adApplication.DisplayName)' does not contain the role '$Role', adding the role"
+
+        $newRole = @{
+          "DisplayName" = $Role
+          "Description" = $Role
+          "Value" = $Role
+          "Id" = [Guid]::NewGuid().ToString()
+          "IsEnabled" = $true
+          "allowedMemberTypes" = @("User", "Application")
+         }
+
+        $adApplication.AppRole += $newRole
+
+        Update-AzADApplication -ObjectId $adApplication.Id -AppRole $adApplication.AppRole
+        Write-Host "Added role '$Role' to Active Directory Application '$($adApplication.DisplayName)'"
+
+        $currentAppRole = $newRole
+    } else {
+        Write-Host "Active Directory Application '$($adApplication.DisplayName)' already contains the role '$Role'"
+        $currentAppRole = $adApplication.AppRole | Where-Object Value -eq $Role
+    }
+
+    $currentRoleAssignments = Get-MgServicePrincipalAppRoleAssignedTo -ServicePrincipalId $adServicePrincipal.Id | Where-Object AppRoleId -eq $currentAppRole.Id 
+    if (-not ($currentRoleAssignments.AppRoleId -contains $currentAppRole.Id)) {
+        $updatedAdServicePrincipal = Get-MgServicePrincipal -ServicePrincipalId $adServicePrincipal.Id
+
+        while (-not ($updatedAdServicePrincipal.AppRoles.Value -contains $Role) -and $counter -lt 10) {
+            Write-Host "Role '$Role' has been added to Active Directory Application '$($adApplication.DisplayName)' but not yet available for use, waiting 10 seconds to retry..."
+            Start-Sleep -Seconds 10
+            $counter++
+            $updatedAdServicePrincipal = Get-MgServicePrincipal -ServicePrincipalId $adServicePrincipal.Id
+        }
+
+        if ($counter -eq 10) {
+            throw "Exhausted the retries, the role '$Role' has been added to Active Directory Application '$($adApplication.DisplayName)' but not yet available for use"
+        }
```

Comment:
Wauw, does it really take that long to add a role assignment programmatically? There's no way to pass in any additional arguments to the `Update-AzADApplication` that would only return if the addition is made successful (so as to avoid waiting here)?
Just thinking out loud, I expect if this script is already used several times and has been adapted over the course of the usage, it would have contain this. 😄

Reply:
Yeah I know, it is frustrating but I have not been able to find a way around this, tried several things but no luck. 
The most frustrating thing is I am migrating this from the `AzureAD` powershell commands where this isn't an issue...